### PR TITLE
devicetree: Update generation/API

### PR DESCRIPTION
**Pull request by galak**
There are cases like having a reserved memory node that will not having
a compatible string and thus will not match a binding.  However we might
set a chosen property like zephyr,sram or zephyr,ipc_shm to such a node.
Thus we should generate defines for such nodes.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I really think we need to discuss this in detail with an issue that describes the impact on the API as discussed on slack.

**Comment by galak** [comment]
> I really think we need to discuss this in detail with an issue that describes the impact on the API as discussed on slack.

what impact will this have on the API?  This isn't related to "status".

**Comment by galak** [comment]
https://github.com/zephyrproject-rtos/zephyr/issues/24773

**Review by ioannisg — APPROVED** [review]
This solves the problem of having new-style DT defines generated for nodes without the compatible property, which seems required in several cases. Looks fine to me, and I don't see any negative impact (but my DT expertise is limited)

File: boards/arm/96b_carbon_nrf51/96b_carbon_nrf51.dts [hunk 1]
```diff
@@ -9,8 +9,7 @@
 
 / {
 	model = "Seeed Studio Carbon nRF51 96board";
-	compatible = "seeed,carbon_nrf51", "nordic,nrf51822-qfac",
-		     "nordic,nrf51822";
+	compatible = "seeed,carbon_nrf51";
```
Comment (pabigot) [inline boards/arm/96b_carbon_nrf51/96b_carbon_nrf51.dts:12]: NAK to this without more motivation.

Compatibles are supposed to be listed from most specific to least specific.  Applications and drivers may inspect this to adjust their behavior within a generic implementation to accommodate the needs of a specific one; or to identify that a specific is also compatible with a generic.  In something I was doing today I would have benefitted greatly from being able to tell I was building on an nRF51 rather than an nrRF52: similar compatibles might have let me do that.

**Review by pabigot — DISMISSED** [review]
I'm OK with the change the PR is proposed to do, but not the secondary one that discards one of the main roles of compatible.

File: boards/arm/96b_carbon_nrf51/96b_carbon_nrf51.dts (hunk 1 above)
Comment (galak) [reply boards/arm/96b_carbon_nrf51/96b_carbon_nrf51.dts:12]: Those compatibles are still available on the SoC node.  Having the same compatible string for 2 different devices isn't really correct.

**Review by pabigot — COMMENTED** [review]
I added these to `/`:

```
       special = "this";
       subspecial {
          bar = "that";
       };
```

The subordinate node was generated, but neither `/special` nor `/subspecial/bar` were generated

Is that what was intended?  Generating the node without its properties seems odd; will this satisfy the need identified in https://github.com/zephyrproject-rtos/zephyr/pull/24460#discussion_r414774455

**Comment by mbolivar-nordic** [comment]
>> I really think we need to discuss this in detail with an issue that describes the impact on the API as discussed on slack.

> what impact will this have on the API? This isn't related to "status".

Here are some ways I find this PR leading to inconsistent or wrong results. Some of them do relate to status.

### DT_HAS_NODE()

This is now returning true for any node that exists in the tree, regardless of whether it has status okay and a matching binding, which contradicts the docstring:

```
 * Tests whether a node identifier refers to a node which:
 *
 * - exists in the devicetree, and
 * - is enabled (has status property "okay"), and
 * - has a matching binding
```

I think that docstring is probably already wrong, since we just need status != "disabled", but that's because of the weird way we deal with status right now, which is maybe a bug now that I look at it.

### DT_NODE_HAS_PROP()

This PR makes it possible to add a node `n` which has no status, and which has no matching binding, and get `DT_HAS_NODE(n) == 1`, but `DT_NODE_HAS_PROP(n, some_random_property_defined_on_n) == 0`, which doesn't make sense to me.

(This is because edtlib refuses to initialize `Node.props` for a node without a matching binding.)

### REG and IRQ APIs

These behave in inconsistent ways between themselves and depending on the status of the node. For example, if some node `foo@0` has no matching binding, these tests can all pass:

```
	zassert_false(DT_NODE_HAS_PROP(DT_PATH(foo_0), reg), "");
	zassert_equal(DT_REG_ADDR(DT_PATH(foo_0)), 0, "");
	zassert_equal(DT_NUM_REGS(DT_PATH(foo_0)), 1, "");
```

In other words, even though the API claims `foo@0` has no reg property, the number of regs can be 1 and you can read a reg address. (This is because edtlib knows what to do with a reg for any node.)

But if you disable the node using status, you can't read its reg anymore, and DT_REG_ADDR is a build error.

Without this PR, things are more consistent. For example, `DT_NODE_HAS_PROP(foo, reg)` returns 0, and using `DT_REG_ADDR` results in a build error, even when the node isn't explicitly disabled, as long as it doesn't have a matching binding.

Similar comments should apply to IRQ properties, though I didn't check explicitly.

**Comment by b0661** [comment]
A while ago a made a proposal to associate a default compatible to special nodes (see #22255). With this you can attach default bindings to such nodes and can use the 'normal' gen_defines.py (and device tree) mechanisms.

**Comment by mbolivar-nordic** [comment]
> A while ago a made a proposal to associate a default compatible to special nodes (see #22255).

This seems like a shorter path to success with the current API

**Comment by pabigot** [comment]
I see the point that this affects existing behavior of `DT_HAS_NODE()`/`DT_HAS_PROP()` but I think that behavior is wrong.  Why isn't `DT_HAS_foo` simply an existence check on the uninterpreted tree?   Node (property) presence should not be conditional on its `status` (description of the property in a binding); that original decision is why we need #19448.

> ... (This is because edtlib refuses to initialize Node.props for a node without a matching binding.)

This is either a showstopper (if edtlib is a reference implementation for devicetree outside Zephyr) or a bug to be fixed.  Has devicetree gone so far towards validation with DTSchema that it will no longer allow nodes that are not documented in a schema?  If so, we need another solution; if not, edtlib should probably make a best guess of the underlying type.  (May relate to https://github.com/zephyrproject-rtos/zephyr/pull/24769#pullrequestreview-402765100)

> But if you disable the node using status, you can't read its reg anymore, and DT_REG_ADDR is a build error.

That also seems like a natural initial solution based on a previous limited interpretation of devicetree (i.e. that only nodes with an "enabled" status are valuable) that should now be reconsidered.

**Comment by galak** [comment]
As I was looking at the EDT changes for this, I'm thinking that we might want to limit this to nodes w/o compatible strings and still require bindings for generation of nodes w/compatible.

**Comment by mbolivar-nordic** [comment]
I think the edtlib change looks sane. It would be good to get the test cases updated with something like this:

- CONFIG_LEGACY_DEVICETREE_MACROS=y for the test application
- verify a node with no binding doesn't have the usual macros available via the legacy macro API

What about the REG/IRQ differences in compile errors when status = enabled vs. disabled in my earlier comment? Do we need to always generate REG macros?

Also I would like @pabigot's input on whether we need to go with the follow-on status changes.

I think the next steps are to add tests verifying DT_HAS_NODE works on nodes with or without bindings, and with status enabled or disabled, and make sure the DT_INST_FOREACH only works on enabled (status okay) nodes. Thoughts?

**Comment by mbolivar-nordic** [comment]
> DT_HAS_NODE works on nodes with or without bindings, and with status enabled or disabled

Though this may take a global search/replace on nodelabel-based device access in samples and drivers.

**Comment by pabigot** [comment]
I agree that the test cases should verify that this does what's expected.  An example based on the use case it's supposed to solve would make this more compelling.

The behavior now is just as I'd described at https://github.com/zephyrproject-rtos/zephyr/pull/24769#pullrequestreview-402765100 except that I'm seeing a few more things generated as well (e.g. compatibles properties for the buttons and LEDs and a few other root nodes like `aliases`).  I understand the concern about generating properties without a binding, but I remain skeptical of the value of generating an empty node.  I guess that can be extended in the future.

@mbolivar not sure what specifically you want my view on.  I would like `DT_HAS_foo` to have the semantics proposed in https://github.com/zephyrproject-rtos/zephyr/pull/24769#issuecomment-621765412 before 2.3 release.  I'm okay with `DT_INST_FOREACH` only working on enabled nodes as long as the documentation is consistent with "enabled" meaning "whatever condition triggers generation of data structures", e.g. `DEFINE_DEVICE*`.  Today that might be "status okay" (which I think is really "status != disabled"), but in the future it must change to potentially support all status values for #19448.

**Comment by galak** [comment]
@mbolivar-nordic are wanting this PR to handle both issues of generating properties for nodes w/o binding matches and the status change?

**Comment by mbolivar-nordic** [comment]
> are wanting this PR to handle both issues of generating properties for nodes w/o binding matches and the status change?

They're interrelated IMO, given the way status can affect REG and IRQ with this change as I detailed in https://github.com/zephyrproject-rtos/zephyr/pull/24769#issuecomment-621621119



> @mbolivar not sure what specifically you want my view on

Sound like you're just wanting it done by release and don't  have a strong view on which PRs should have things done in. OK.

File: scripts/dts/gen_defines.py [hunk 2]
```diff
@@ -59,9 +59,6 @@ def main():
             if not node.enabled:
                 out_comment("No node macros: node is disabled")
```
Comment (mbolivar-nordic) [inline scripts/dts/gen_defines.py:60]: So if we're moving to https://github.com/zephyrproject-rtos/zephyr/pull/25002, I think this should be removed too, right? We had discussed being able to read any property we have semantics for (at least 'reg' and 'interrupts') regardless of status.

Comment (mbolivar-nordic) [reply scripts/dts/gen_defines.py:60]: I also think things like `compat2enabled` in the EDT object may need changes, since we're changing the semantics of "enabled" in C to be `status == "okay"` instead of `status != "disabled"` and adding macros like DT_HAS_COMPAT_STATUS_OKAY in https://github.com/zephyrproject-rtos/zephyr/pull/25002. So it seems like we should be doing `compat2statusokay` or something instead.

In general I'm having trouble keeping all the pull requests together and I don't see enough Python changes to accompany all the C changes. What am I missing?

**Comment by b0661** [comment]
```
edtlib: populate standard properties for nodes without bindings

If a devicetree node doesn't have a matching binding we will at least
populate a common standard set ...:

	compatible
	status
	reg
	reg-names
	label
	interrupt
	interrupts-extended
	interrupt-names
	interrupt-controller

This allows us to handle cases like memory nodes ...
```

Why not just assign a default.yaml binding to active nodes without a binding. The binding can contain any property of above or just include base.yaml.

The explicit addition of default properties in the current solution does not scale well and also makes edtlib.py less generic.

**Comment by galak** [comment]
> Why not just assign a default.yaml binding to active nodes without a binding. The binding can contain any property of above or just include base.yaml.
> 
> The explicit addition of default properties in the current solution does not scale well and also makes edtlib.py less generic.

I was torn about this, since it makes an assumption about there being a specific file in one of the binding dirs.

File: include/devicetree.h [hunk 3]
```diff
@@ -880,6 +880,32 @@
...
+ *
+ *     DT_NODE_HAS_STATUS(DT_NODELABEL(n), disabled)
+ *
+ * @param node_id node identifier
+ * @param status status value to check for
+ * @return 1 if the node's status property matches status,
+ *         0 otherwise.
+ */
+#define DT_NODE_HAS_STATUS(node_id, status) \
+	IS_ENABLED(DT_CAT(node_id, _STATUS_##status))
```
Comment (mbolivar-nordic) [inline include/devicetree.h:907]: my approach in #25009 works even when status is a macro that needs to be expanded again, unlike this one

File: include/devicetree.h [hunk 4]
```diff
@@ -898,7 +924,9 @@
  * @return 1 if the node identifier refers to a usable node,
  *         0 otherwise.
  */
-#define DT_HAS_NODE_STATUS_OKAY(node_id) IS_ENABLED(DT_CAT(node_id, _EXISTS))
+#define DT_HAS_NODE_STATUS_OKAY(node_id) \
+	UTIL_AND(IS_ENABLED(DT_CAT(node_id, _EXISTS)), \
```
Comment (mbolivar-nordic) [inline include/devicetree.h:928]: It's impossible to have a node that has status okay but doesn't exist as far as I can tell. That's why #25009 doesn't have this check. Am I missing something?

File: tests/lib/devicetree/src/main.c [hunk 5]
```diff
@@ -200,6 +200,27 @@ static void test_has_compat(void)
...
+	zassert_equal(DT_NODE_HAS_STATUS(DT_NODELABEL(test_gpio_1), disabled),
+		      0, "vnd,gpio not disabled");
+
+	zassert_equal(DT_NODE_HAS_STATUS(DT_NODELABEL(test_no_status), okay),
+		      1, "vnd,gpio okay");
+	zassert_equal(DT_NODE_HAS_STATUS(DT_NODELABEL(test_no_status), disabled),
+		      0, "vnd,gpio not disabled");
+
+/* Disabled for now until we generate something for non-"enabled" nodes */
+#if 0
```
Comment (mbolivar-nordic) [inline tests/lib/devicetree/src/main.c:216]: Leaving a comment here since we've merged ifdeffed out tests before, to make sure it doesn't happen this time

File: include/devicetree.h (hunk 4 above)
Comment (galak) [reply include/devicetree.h:928]: Nope, you're correct we don't need the extra check.

File: include/devicetree.h (hunk 3 above)
Comment (galak) [reply include/devicetree.h]: Changed to be like your version.

File: include/devicetree.h (hunk 4 above)
Comment (galak) [reply include/devicetree.h]: Changed.

File: scripts/dts/gen_defines.py (hunk 2 above)
Comment (galak) [reply scripts/dts/gen_defines.py:60]: I reworked how enabled was defined.  If we want to leave that alone we can create a new 'available' which would be similar to the linux concept https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/of/base.c#n581

**Comment by b0661** [comment]
>>   Why not just assign a default.yaml binding to active nodes without a binding. The binding can contain any property of above or just include base.yaml.
>>  The explicit addition of default properties in the current solution does not scale well and also makes edtlib.py less generic.

> I was torn about this, since it makes an assumption about there being a specific file in one of the binding dirs.

If the specific binding file default.yaml is not found it does not hurt. If it is found it is available on purpose (like e.g. base.yaml) and the default properties will be generated.

With the current solution you get the opposite situation - every user of edtlib.py gets the default properties generated. This may not be a problem within Zephyr but usability outside of Zephyr may be reduced.

Using default.yaml gives a choice and does not reduce usability in any way.

**Comment by mbolivar-nordic** [comment]
> Using default.yaml gives a choice and does not reduce usability in any way.

I've chosen a different strategy with a similar effect in #25009. Please have a look.

**Comment by galak** [comment]
I get the desire @b0661 and would say we can add some way to optionally extend the base set of properties if someone desires that.  However, while I'm not super happy about hard coding a list of properties into EDT, I think its the most reasonable tradeoff for the set of properties that are defined as part of the base DTS spec.
